src/ts/client/grapesjs/keymaps.ts

```diff
@@ -0,0 +1,111 @@
+import {Editor, PluginOptions} from 'grapesjs'
+import {cmdPublish} from './PublicationUi'
+import {cmdOpenFonts} from '@silexlabs/grapesjs-fonts'
+import {cmdToggleBlocks, cmdToggleLayers, cmdToggleNotifications, cmdToggleSymbols} from './index'
+import {cmdTogglePages} from './page-panel'
+import {cmdOpenSettings} from './settings'
+import {isTextOrInputField, selectBody} from '../utils'
+import {PublishableEditor} from './PublicationManager'
+
+export const cmdSelectBody = 'select-body'
+export let prefixKey = 'Shift'
+export const defaultKms = {
+  kmOpenSettings: 'Alt+S',
+  kmOpenPublish: 'Alt+P',
+  kmOpenFonts: 'Alt+F',
+  kmPreviewMode: 'tab',
+  kmLayers: prefixKey + '+L',
+  kmBlocks: prefixKey + '+A',
+  kmNotifications: prefixKey + '+N',
+  kmPages: prefixKey + '+P',
+  kmSymbols: prefixKey + '+S',
+  kmClosePanel: 'Escape'
+}
+
+// Utility functions
+
+const toggleCommand = (editor: Editor, name: string): void => {
+  const cmd = editor.Commands
+
+  if (!cmd.isActive(name)) {
+    resetPanel(editor)
+    cmd.run(name)
+  } else {
+    cmd.stop(name)
+  }
+}
+
+const resetPanel = (editor: Editor): void => {
+  const panels: string[] = [
+    cmdToggleBlocks,
+    cmdToggleLayers,
+    cmdToggleNotifications,
+    cmdToggleSymbols,
+    cmdTogglePages,
+    cmdOpenSettings,
+    cmdOpenFonts
+  ]
+  panels.forEach(p => editor.Commands.stop(p))
+}
+
+/**
+ * Escapes the current context in this order : modal, Publish dialog, left panel.
+ * @param editor The editor.
+ */
+const escapeContext = (editor: Editor): void => {
+  const publishDialog = (editor as PublishableEditor).PublicationManager.dialog
+
+  if (editor.Modal.isOpen()) {
+    console.log('closing modal')
+    editor.Modal.close()
+  } else if (publishDialog && publishDialog.isOpen) {
+    publishDialog.closeDialog()
+  } else {
+    resetPanel(editor)
+  }
+}
+
+// Main part
+
+export const keymapsPlugin = (editor: Editor, opts: PluginOptions): void => {
+  // Commands
+  editor.Commands.add(cmdSelectBody, selectBody)
+
+  if (opts.disableKeymaps) return
+  if (opts.prefixKey) prefixKey = opts.prefixKey
+
+  const km = editor.Keymaps
+
+  // Panels
+  km.add('general:open-settings', defaultKms.kmOpenSettings.toLowerCase(), editor => toggleCommand(editor, cmdOpenSettings))
```

Comment:
it would be easier to understand if defaultKms is an array of objects with [keymap::add params](https://grapesjs.com/docs/api/keymaps.html#add) id, keys, handler and opts attributes

then you could loop over tye array and save the code bellow?

Reply:
or maybe keep defaultKms an object, with the IDs as keys and keys, handler and opts as attributes

Reply:
I thought about this, but if you look at the content of ``defaultKms``, you will see that I put majuscules. It is because I use them for keyboard hints, but you need to lower the case before declaring the keymaps.
What I could do is make it an object with all the properties needed, and when I need a "cleaner" display, I can add a conversion function into ``utils.ts``.

Reply:
(Done)